### ISSUE TITLE
About training parameters for WRN

### ISSUE THREAD

**Issue by SnowyJune973** [mboudiaf/TIM#6]
Thank you for releasing code for such a creative method! However I've faced some problems when reproducing the results.

From what I've known, TIM is one of the most s-o-t-a few-shot methods using WRN as backbone at present. But the pretraining strategy shown in the code seems problematic, as I could only train a model that gets a 34.99% accuracy in a 16-way 1-shot task on  miniImageNet, comparing to ~44% with the strategy used in SIB or EPNet. It has been pointed out that the quality of pretrained backbone will significantly influence the performance of the method, and since TIM outperforms SIB and EPNet a lot with no extra fine-tuning stage, this becomes puzzling.

This the training strategy for (miniImageNet, WRN) I found in your code:
inital LR=0.1, optimizer: SGD w/ nesterov momentum=0.9 , weight decay=1e-4
N_epoch=90
LR schedule: multistep, LR*=0.1@epoch 45&67
Label smoothing 0.1
Data augment: Color jitter

Did I miss something? If not, could you tell me why you didn't adopt a better pretrained backbone to further improve the results? Thanks a lot.

**Comment by mboudiaf** [comment]
Hi ! 
Thank you very much for your interest and for raising this issue. I've tried running 10'000 16-ways -1-shot tasks using the models I had trained (the ones available for download) and I get the following convergence plot for TIM-GD: 

![image](https://user-images.githubusercontent.com/40210629/108272768-ad0ba800-7140-11eb-9087-ddc3271f7fd4.png)

with a final result just below 49%, and a baseline (i.e SimpleShot method) around 39%, which makes more sense. I will investigate further on the training script, and run everything from scratch to make sure no mistake was introduced in my latest commits. I will get back to you soon.

**UPDATE**  : It seems in my latest commit I introduced a duplicate scheduler.step() in both main.py and trainer.py, which caused the learning rate to decay twice as fast as it should, degrading the results. Using the "bugged" coded, I was still able to get 44% in your setting, and fixing the bug I could obtain 48% + after training. I have pushed my changes. Please could you pull and see if that solves your problem ? Thank you :) 

Malik

**Comment by SnowyJune973** [comment]
> Hi !
> Thank you very much for your interest and for raising this issue. I've tried running 10'000 16-ways -1-shot tasks using the models I had trained (the ones available for download) and I get the following convergence plot for TIM-GD:
> 
> ![image](https://user-images.githubusercontent.com/40210629/108272768-ad0ba800-7140-11eb-9087-ddc3271f7fd4.png)
> 
> with a final result just below 49%, and a baseline (i.e SimpleShot method) around 39%, which makes more sense. I will investigate further on the training script, and run everything from scratch to make sure no mistake was introduced in my latest commits. I will get back to you soon.
> 
> **UPDATE** : It seems in my latest commit I introduced a duplicate scheduler.step() in both main.py and trainer.py, which caused the learning rate to decay twice as fast as it should, degrading the results. Using the "bugged" coded, I was still able to get 44% in your setting, and fixing the bug I could obtain 48% + after training. I have pushed my changes. Please could you pull and see if that solves your problem ? Thank you :)
> 
> Malik

Thank you for your efforts! 
It was my fault that I didn't mention further details on how I obtained the 39.44% result. I didn't use TIM-GD or other TIM-* methods to do this, instead I trained a classifier with WRN as the backbone and a FC layer attached to it, using the training parameters adopted in this project, so I think the "bugged" code will not affect the result I got (my codes are based on pretrain.py and model/models/classifier.py in [FEAT](https://github.com/Sha-Lab/FEAT)) . 
It should be fine if the configuration in the scripts/train/wideres.sh and src/trainer.py is nothing wrong. Besides, I noticed you introduced a MixUp augmentation in the training process of trainer.py, but it seems to be deprecated in recent versions. Could you tell me if you activated it when getting the results shown on the paper? That may be a crucial factor. Thanks :)

**Comment by mboudiaf** [comment]
Hi,

1) I recommend you use the code from this repo to pre-train your model if you want to obtain results for TIM. From what I see, the code in pretrain.py from the FEAT code is not using a standard cross-entropy training as we do, but an episodic one. In this case, using the same hyperparameters as the ones I am using does not really make sense. Am I missing something ?
2) I have left the possibility to use mixup for future attempts, but the results presented in our paper does not include mixup augmentation during pre-training.

Best,
Malik
